### DISCUSSION TITLE
Dataset CRS vs. Project CRS

### DISCUSSION THREAD

**Discussion by martimmas** [qgreenland-workshop-2023-researcher/qgreenland-workshop-2023-researcher.github.io#152] [Help Requests] [answered]
One question/problem we ran into: What kind of behaviour would you expect in QGreenland/QGIS when loading a global dataset and the project's CRS is set to a Polar stereographic projection?

**Comment by MattF-NSIDC** [comment]
Hey Martim, thanks for asking! We'll actually touch on this a little bit during tomorrow's morning module :)

When you load a dataset into QGIS and its projection doesn't match the project projection (for QGreenland, that's `EPSG:3413`), QGIS will do what's called "on-the-fly" reprojection, where the data is reprojected under the hood (without modifying the underlying data file) so it can be viewed alongside other layers in the project projection. This can be quite slow as you pan and zoom around the map, especially when very large datasets are reprojected on-the-fly.

If this post answers your question, please use the "Mark as answer" button for future readers!

**Comment by martimmas** [comment]
Hi Matt,

Thanks for the reply! I suspected QGIS was doing that when I loaded the dataset, but what got me curious was this specific case when a global dataset is reprojected to a polar stereographic CRS. What happens at the boundaries, and how does QGIS/gdal/whatever software you end up using deal with that? Would you get an "infinite" border, some weird distortion, or is there a "cutoff latitude" to prevent anything weird coming from the opposite pole? 

Not sure if I was clear with my explantion, but let me know otherwise! And thanks a lot once again for this workshop!

**Reply by MattF-NSIDC** [reply, marked as answer]
That makes sense, thanks for clarifying :)

To produce these examples, I used the Natural Earth 2 raster dataset.

In QGIS, the data that is reprojected on-the-fly is limited to the viewport area. The farther you zoom out, the longer the reprojection will take. This is the farthest I could get QGIS to zoom out:

![image](https://github.com/qgreenland-workshop-2023-researcher/qgreenland-workshop-2023-researcher.github.io/assets/18539827/a8c00d48-0ba6-4195-9b06-888d9c4c3d8e)

Here's what `gdalwarp` does:

```
$ gdalwarp -t_srs EPSG:3413 NE2_50M_SR_W.tif NE2_50M_SR_W_polarstereo.tif 
Creating output file that is 8536P x 8540L.
Processing NE2_50M_SR_W.tif [1/1] : 0ERROR 1: Too many points (529 out of 529) failed to transform, unable to compute output bounds.
Warning 1: Unable to compute source region for output window 0,0,8536,8540, skipping.
...10...20...30...40...50...60...70...80...90...100 - done.
```

The input file was 167M, and the output file is 209M. Opening the file in QGIS or an image viewer that supports TIF files shows only black. Using `xarray` to open the output data and `numpy` to get unique values, we can see that all values came out as `0`:

```
import xarray as xra
import numpy as np

ds = xra.open_dataset('NE2_50M_SR_W_polarstereo.tif')
list(np.unique(ds.band_data.to_numpy()))
```

this code produces: `[0.0]`. GDAL has failed to do what we asked because it couldn't figure out what to do at the edges!

If you need to do a projection like this, it's key to do a subset/clip operation in advance with your target projection in mind.  For example, if you want to go from a global CRS to a north polar stereographic (where the south pole can't be represented), you might subset to your area of interest first, then reproject. Or, `gdalwarp` is capable of doing this with the `-te` (target extent) argument in the same command as the warp operation.

When QGreenland subsets this basemap, we use the `gdalwarp` `-te` (target extent) argument to define the output extent we want to limit to (in the target projected coordinate reference system): [source code](https://github.com/nsidc/qgreenland/blob/3bcb656e3d2cf433a213083b48b7e2206f7a7403/qgreenland/config/layers/Basemaps/background.py#L43-L44)

**Reply by martimmas** [reply]
That's an amazing and very comprehensive answer, Matt! Thanks a lot!
